hudi-common/src/main/java/org/apache/hudi/BaseHoodieTableFileIndex.java

```diff
@@ -304,6 +304,18 @@ private void doRefresh() {
                   .orElse(Collections.emptyList())
               )
           );
+    } else if (tableType.equals(HoodieTableType.MERGE_ON_READ) && queryType.equals(HoodieTableQueryType.READ_OPTIMIZED)) {
+      cachedAllInputFileSlices = partitionFiles.keySet().stream()
+          .collect(Collectors.toMap(
+                  Function.identity(),
+                  partitionPath ->
+                      queryInstant.map(instant ->
+                              fileSystemView.getLatestMergedFileSlicesWithOnlyBaseFileBeforeOrOn(partitionPath.path, queryInstant.get())
```

Comment:
Basically, you're fixing the problem where the compaction is in progress and the file slice falling into the compaction under construction should not be included until it is committed.  Shall we fix the `getLatestFileSlicesBeforeOrOn` instead of adding a new API here?